c/meterpreter/source/extensions/priv/service.c

```diff
@@ -172,3 +172,47 @@ DWORD service_destroy( char * cpName )
 
 	return dwResult;
 }
+
+/*
+ * Query service state.
+ */
+DWORD query_service_status( char* cpName, DWORD* dwState )
+{
+	DWORD dwResult  = ERROR_SUCCESS;
+	HANDLE hManager = NULL;
+	HANDLE hService = NULL;
+	DWORD dwBytes;
+	SERVICE_STATUS_PROCESS procInfo;
+
+	do
+	{
+		if( !cpName )
+			BREAK_WITH_ERROR( "[SERVICE] query_service_status. cpName is NULL", ERROR_INVALID_HANDLE );
+
+		hManager = OpenSCManagerA( NULL, NULL, SC_MANAGER_CONNECT);
+		if( !hManager )
+			BREAK_ON_ERROR( "[SERVICE] query_service_status. OpenSCManagerA failed" );
+
+		hService = OpenServiceA( hManager, cpName, SERVICE_QUERY_STATUS);
+		if( !hService )
+			BREAK_ON_ERROR( "[SERVICE] query_service_status. OpenServiceA failed" );
+
+		if (!QueryServiceStatusEx(hService, SC_STATUS_PROCESS_INFO, (LPBYTE)&procInfo, sizeof(procInfo), &dwBytes)) {
+			BREAK_ON_ERROR("[SERVICE] query_service_status. QueryServiceStatusEx failed");
+		}
+		else {
+			*dwState = procInfo.dwCurrentState;
+		}
+
+	} while( 0 );
+
+	if( hService )
+		CloseServiceHandle( hService ); 
+
+	if( hManager )
+		CloseServiceHandle( hManager ); 
+
+	SetLastError( dwResult );
```

Comment:
Umm this will only ever be `ERROR_SUCCESS` since we only ever initialize this value and we never change its value throughout this code. We will want to adjust this code so that we set this to an error code if the handles can't be open.

Reply:
Actually [`BREAK_ON_ERROR` ](https://github.com/rapid7/metasploit-payloads/blob/cc3459772878c73ba692ea561a6fdbc44d56a809/c/meterpreter/source/common/common.h#L90) will set `dwResult`. Having said that, I'm not sure it's necessary to pass it to `SetLastError` here.

Reply:
`dwResult` actually gets updated by `BREAK_ON_ERROR` macro, used in function.

Reply:
@NtAlexio2 Not sure why this was marked as resolved. Spencer's point about using `SetLastError` here still stands.